src/main/java/org/apache/sysml/runtime/instructions/gpu/context/JCudaKernels.java

```diff
@@ -18,142 +18,43 @@
  */
 package org.apache.sysml.runtime.instructions.gpu.context;
 
-import static jcuda.driver.JCudaDriver.cuLaunchKernel;
-import static jcuda.driver.JCudaDriver.cuModuleGetFunction;
-import static jcuda.driver.JCudaDriver.cuModuleLoadDataEx;
+import jcuda.Sizeof;
 
-import java.io.ByteArrayOutputStream;
-import java.io.IOException;
-import java.io.InputStream;
-import java.util.HashMap;
-
-import org.apache.sysml.runtime.DMLRuntimeException;
-import org.apache.sysml.runtime.io.IOUtilFunctions;
+// This indirection prepares SystemML's GPU backend to allow
+// for arbitrary datatype
+public abstract class JCudaKernels {
+	private static JCudaKernels singletonObj = new DoublePrecisionKernels();
```

Comment:
I'd prefer no singleton objects, specially when you get it from an abstract class. I prefer factories for this sort of stuff. Granted that it adds more cruft. 
An alternative way to do this is to add the `sizeOfDatatype()` and `cudnnDataType()` methods to jcudakernels, or more aptly to `GPUContext` (since all cublas, cusparse, cusolve, cudnn library calls need to have that data type, not just the custom kernels). In the GPUContext, you could control which data type to select via an enumeration. 
I recommend an enumeration here, because (a) the types are limited - float, double, short... (b) what you plan to do with this information is very limited (`cudnnDataType` and `sizeOfDatatype`).

Also, the decision to make it single vs double precision should not be embedded this deep down in the code. 
it should be controlled by a systemml property (from the SystemML-config.xml file) or via a command line parameter.